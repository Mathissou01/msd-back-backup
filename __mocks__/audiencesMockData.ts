import { GetAudiencesByContractIdDocument } from "../src/graphql/codegen/generated-types";

export const getNoAudienceMockedData = [
  {
    request: {
      query: GetAudiencesByContractIdDocument,
      variables: {
        filters: {
          contract: {
            id: {
              eq: undefined,
            },
          },
          isActive: {
            eq: true,
          },
        },
      },
    },
    result: {
      data: {
        audiences: {
          data: [],
        },
      },
    },
  },
];

export const getSingleAudienceMockedData = [
  {
    request: {
      query: GetAudiencesByContractIdDocument,
      variables: {
        filters: {
          contract: {
            id: {
              eq: undefined,
            },
          },
          isActive: {
            eq: true,
          },
        },
      },
    },
    result: {
      data: {
        audiences: {
          data: [
            {
              id: 1,
              attributes: {
                isActive: true,
                type: "Particuliers",
                __typename: "Audience",
              },
            },
          ],
        },
      },
    },
  },
];

export const getTwoAudiencesMockedData = [
  {
    request: {
      query: GetAudiencesByContractIdDocument,
      variables: {
        filters: {
          contract: {
            id: {
              eq: undefined,
            },
          },
          isActive: {
            eq: true,
          },
        },
      },
    },
    result: {
      data: {
        audiences: {
          data: [
            {
              id: 1,
              attributes: {
                isActive: true,
                type: "Particuliers",
                __typename: "Audience",
              },
            },
            {
              id: 2,
              attributes: {
                isActive: true,
                type: "Collectifs",
                __typename: "Audience",
              },
            },
          ],
        },
      },
    },
  },
];

export const getAllAudiencesMockedData = [
  {
    request: {
      query: GetAudiencesByContractIdDocument,
      variables: {
        filters: {
          contract: {
            id: {
              eq: undefined,
            },
          },
          isActive: {
            eq: true,
          },
        },
      },
    },
    result: {
      data: {
        audiences: {
          data: [
            {
              id: 1,
              attributes: {
                isActive: true,
                type: "Particuliers",
                __typename: "Audience",
              },
            },
            {
              id: 2,
              attributes: {
                isActive: true,
                type: "Collectifs",
                __typename: "Audience",
              },
            },
            {
              id: 3,
              attributes: {
                isActive: true,
                type: "Professionnels",
                __typename: "Audience",
              },
            },
          ],
        },
      },
    },
  },
];
