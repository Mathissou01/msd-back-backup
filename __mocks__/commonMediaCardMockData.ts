import { GetFilesPaginationByFolderIdDocument } from "../src/graphql/codegen/generated-types";

export const defaultMockData = [
  {
    request: {
      query: GetFilesPaginationByFolderIdDocument,
      filters: {
        folder: {
          id: {
            eq: 2,
          },
        },
      },
      pagination: {
        page: 1,
        pageSize: 10,
      },
      sort: "createdAt:desc",
    },
    result: {
      data: {
        uploadFiles: {
          data: [
            {
              id: "124",
              attributes: {
                name: "La Bruja.jpeg",
                mime: "image/jpeg",
                size: 124.35,
                width: 1280,
                height: 853,
                createdAt: "2023-01-26T10:34:26.058Z",
                url: "https://stomsdmediadev.blob.core.windows.net/msd-media-dev/assets-local-sdial/La%20Bruja.jpeg",
                ext: ".jpeg",
              },
            },
            {
              id: "122",
              attributes: {
                name: "Banana Mama Rooftop & Kitchen 5.jpeg",
                mime: "image/jpeg",
                size: 68.84,
                width: 960,
                height: 960,
                createdAt: "2023-01-26T10:30:22.701Z",
                url: "https://stomsdmediadev.blob.core.windows.net/msd-media-dev/assets-local-sdial/Banana%20Mama%20Rooftop%20%26%20Kitchen%205.jpeg",
                ext: ".jpeg",
              },
            },
            {
              id: "121",
              attributes: {
                name: "Sichuan Hot Pot.jpeg",
                mime: "image/jpeg",
                size: 2233.13,
                width: 6962,
                height: 4615,
                createdAt: "2023-01-18T17:39:30.518Z",
                url: "https://stomsdmediadev.blob.core.windows.net/msd-media-dev/assets-local-sdial/Sichuan%20Hot%20Pot.jpeg",
                ext: ".jpeg",
              },
            },
            {
              id: "120",
              attributes: {
                name: "mahbod-akhzami-kg40JrEsCvY-unsplash.jpg",
                mime: "image/jpeg",
                size: 1788.48,
                width: 6000,
                height: 4000,
                createdAt: "2023-01-18T17:39:29.900Z",
                url: "https://stomsdmediadev.blob.core.windows.net/msd-media-dev/assets-local-sdial/mahbod-akhzami-kg40JrEsCvY-unsplash.jpg",
                ext: ".jpg",
              },
            },
            {
              id: "119",
              attributes: {
                name: "Paella in Valencia.jpeg",
                mime: "image/jpeg",
                size: 1026.93,
                width: 3851,
                height: 2528,
                createdAt: "2023-01-18T17:39:29.720Z",
                url: "https://stomsdmediadev.blob.core.windows.net/msd-media-dev/assets-local-sdial/Paella%20in%20Valencia.jpeg",
                ext: ".jpeg",
              },
            },
            {
              id: "118",
              attributes: {
                name: "Rusconi's American Kitchen 3.jpeg",
                mime: "image/jpeg",
                size: 209.81,
                width: 1999,
                height: 1333,
                createdAt: "2023-01-18T17:39:29.502Z",
                url: "https://stomsdmediadev.blob.core.windows.net/msd-media-dev/assets-local-sdial/Rusconi%27s%20American%20Kitchen%203.jpeg",
                ext: ".jpeg",
              },
            },
            {
              id: "117",
              attributes: {
                name: "Quince 5.jpeg",
                mime: "image/jpeg",
                size: 197.32,
                width: 2000,
                height: 1500,
                createdAt: "2023-01-18T17:39:29.310Z",
                url: "https://stomsdmediadev.blob.core.windows.net/msd-media-dev/assets-local-sdial/Quince%205.jpeg",
                ext: ".jpeg",
              },
            },
            {
              id: "116",
              attributes: {
                name: "Rusconi's American Kitchen 2.jpeg",
                mime: "image/jpeg",
                size: 72.49,
                width: 1000,
                height: 1500,
                createdAt: "2023-01-18T17:39:29.221Z",
                url: "https://stomsdmediadev.blob.core.windows.net/msd-media-dev/assets-local-sdial/Rusconi%27s%20American%20Kitchen%202.jpeg",
                ext: ".jpeg",
              },
            },
            {
              id: "115",
              attributes: {
                name: "Pur' - Jean-Fran√ßois Rouquette 3.jpeg",
                mime: "image/jpeg",
                size: 63.2,
                width: 1280,
                height: 853,
                createdAt: "2023-01-18T17:39:29.039Z",
                url: "https://stomsdmediadev.blob.core.windows.net/msd-media-dev/assets-local-sdial/Pur%27%20-%20Jean-Fran%C3%A7ois%20Rouquette%203.jpeg",
                ext: ".jpeg",
              },
            },
            {
              id: "114",
              attributes: {
                name: "muneeb-muhamed-kPXT9HA4LOk-unsplash.jpg",
                mime: "image/jpeg",
                size: 516.64,
                width: 3456,
                height: 4608,
                createdAt: "2023-01-18T17:39:28.922Z",
                url: "https://stomsdmediadev.blob.core.windows.net/msd-media-dev/assets-local-sdial/muneeb-muhamed-kPXT9HA4LOk-unsplash.jpg",
                ext: ".jpg",
              },
            },
          ],
          meta: {
            pagination: {
              total: 122,
              pageSize: 10,
              pageCount: 13,
              page: 1,
            },
          },
        },
      },
    },
  },
];
