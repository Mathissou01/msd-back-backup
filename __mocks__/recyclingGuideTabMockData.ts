import { GetRecyclingBlockTabDocument } from "../src/graphql/codegen/generated-types";

export const defaultMockData = [
  {
    request: {
      query: GetRecyclingBlockTabDocument,
      variables: {
        contractId: "0",
      },
    },
    result: {
      data: {
        contractCustomizations: {
          data: [
            {
              attributes: {
                homepage: {
                  data: {
                    attributes: {
                      recyclingGuideBlock: {
                        data: {
                          id: "1",
                          attributes: {
                            titleContent: "Titre modifié",
                            subtitleContent:
                              "Trouver les consignes de tri dans ma ville",
                            recyclingGuideDisplayContent:
                              "Lampe, écran, épluchure",
                            tags: [],
                          },
                        },
                      },
                    },
                  },
                },
              },
            },
          ],
        },
      },
    },
  },
];
