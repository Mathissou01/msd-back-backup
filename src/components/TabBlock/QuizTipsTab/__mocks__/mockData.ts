import { GetQuizAndTipsBlockDocument } from "../../../../graphql/codegen/generated-types";

export const defaultMockData = [
  {
    request: {
      query: GetQuizAndTipsBlockDocument,
      variables: {
        contractId: "1",
      },
    },
    result: {
      data: {
        contractCustomizations: {
          data: [
            {
              attributes: {
                homepage: {
                  data: {
                    attributes: {
                      quizAndTipsBlock: {
                        data: {
                          id: "1",
                          attributes: {
                            title: "Titre modifi√©",
                            displayBlock: false,
                            displayQuiz: false,
                            quiz: {
                              data: {
                                id: "2",
                                attributes: {
                                  title: "Nom du quiz",
                                  publishedAt: "2022-11-20T02:34:14.535Z",
                                },
                              },
                            },
                            displayTips: false,
                            tips: {
                              data: [],
                            },
                          },
                        },
                      },
                    },
                  },
                },
              },
            },
          ],
        },
        services: {
          data: [
            {
              attributes: {
                serviceInstance: [
                  {
                    __typename: "ComponentMsdEditorial",
                    id: "4",
                    editorialServices: {
                      data: [
                        {
                          attributes: {
                            subServiceInstance: [
                              {
                                __typename: "ComponentEditoQuizzesSubService",
                                isActivated: true,
                                quizzes: {
                                  data: [
                                    {
                                      id: "2",
                                      attributes: {
                                        title: "Nom du quiz",
                                        publishedAt: "2022-11-20T02:34:14.535Z",
                                      },
                                    },
                                    {
                                      id: "1",
                                      attributes: {
                                        title: "truc",
                                        publishedAt: "2022-11-18T10:16:08.936Z",
                                      },
                                    },
                                  ],
                                },
                              },
                            ],
                          },
                        },
                        {
                          attributes: {
                            subServiceInstance: [
                              {
                                __typename: "ComponentEditoTipsSubService",
                                isActivated: true,
                                tips: {
                                  data: [
                                    {
                                      id: "5",
                                      attributes: {
                                        title: "nah",
                                        publishedAt: "2022-11-19T00:14:37.403Z",
                                      },
                                    },
                                    {
                                      id: "2",
                                      attributes: {
                                        title: "Titre de l'astuce 1",
                                        publishedAt: "2022-11-19T00:14:08.232Z",
                                      },
                                    },
                                    {
                                      id: "1",
                                      attributes: {
                                        title: "Titre de l'astuce 2",
                                        publishedAt: "2022-11-18T10:17:11.158Z",
                                      },
                                    },
                                    {
                                      id: "3",
                                      attributes: {
                                        title: "Titre de l'astuce 3",
                                        publishedAt: "2022-11-19T00:14:20.490Z",
                                      },
                                    },
                                    {
                                      id: "4",
                                      attributes: {
                                        title: "Titre de l'astuce 4",
                                        publishedAt: "2022-11-19T00:14:30.273Z",
                                      },
                                    },
                                  ],
                                },
                              },
                            ],
                          },
                        },
                      ],
                    },
                  },
                ],
              },
            },
          ],
        },
      },
    },
  },
];

export const errorMockData = [
  {
    request: {
      query: GetQuizAndTipsBlockDocument,
      variables: {
        contractId: "1",
      },
    },
    error: new Error("An error occured"),
  },
];
