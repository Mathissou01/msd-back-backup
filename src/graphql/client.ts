import {
  ApolloClient,
  InMemoryCache,
  NormalizedCacheObject,
} from "@apollo/client";

let client: ApolloClient<NormalizedCacheObject>;

if (process.env.NEXT_PUBLIC_MOCK === "true") {
  /* With .env variable NEXT_PUBLIC_MOCK=true, return mock graphql client with fake response data */
  client = (await import("./mockedClient")).default(1000);
} else {
  client = await new ApolloClient({
    uri: `${process.env.NEXT_PUBLIC_API_URL}/graphql`,
    cache: new InMemoryCache({
      typePolicies: {
        UploadFolder: {
          merge: true,
        },
        UploadFile: {
          merge: true,
        },
        Contract: {
          merge: true,
        },
        ContractCustomization: {
          merge: true,
        },
        New: {
          merge: true,
        },
        Tip: {
          merge: true,
        },
        RecyclingGuideService: {
          merge: true,
        },
        WasteForm: {
          merge: true,
        },
        WasteFamily: {
          merge: true,
        },
        Flow: {
          merge: true,
        },
        FlowColor: {
          merge: true,
        },
        DropOffMap: {
          merge: true,
        },
        Request: {
          merge: true,
        },
        RequestAggregate: {
          merge: true,
        },
      },
    }),
  });
}

export default client;
