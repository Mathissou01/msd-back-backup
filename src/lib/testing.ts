import { prettyDOM } from "@testing-library/react";

export function removeUnstableHtmlPropertiesForTesting(
  htmlElement: HTMLElement,
  attributeName: string,
) {
  const domHTML = prettyDOM(htmlElement, Infinity).toString();
  if (!domHTML) return undefined;
  const regex = new RegExp(`${attributeName}(.*)"(.*)"`, "g");
  return domHTML.replaceAll(regex, "");
}
