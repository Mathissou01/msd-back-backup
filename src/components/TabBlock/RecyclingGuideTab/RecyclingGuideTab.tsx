import { FormProvider, useForm } from "react-hook-form";
import { FieldValues } from "react-hook-form/dist/types/fields";
import React, { useState, useCallback, useEffect, useRef } from "react";
import {
  GetRecyclingBlockTabDocument,
  useGetRecyclingBlockTabQuery,
  useUpdateRecyclingGuideTabMutation,
} from "../../../graphql/codegen/generated-types";
import { FocusFirstElement } from "../../../lib/utilities";
import { extractRecyclingGuideBlock } from "../../../lib/graphql-data";
import CommonButton from "../../Common/CommonButton/CommonButton";
import CommonSpinner from "../../Common/CommonSpinner/CommonSpinner";
import FormInput from "../../Form/FormInput/FormInput";
import "./recycling-guide-tab.scss";
import { useContract } from "../../../hooks/useContract";

interface IRecyclingGuideBlock {
  id: string;
  titleContent: string;
  subtitleContent: string;
  recyclingGuideDisplayContent: string;
}

export default function RecyclingGuideTab() {
  /* Static Data */
  const formLabels = {
    title: "Guide du Tri",
    maxCharactersLabel: "caractères maximum",
    titleContent: "Titre",
    subtitleContent: "Sous titre",
    recyclingGuideDisplayContent:
      "Texte affiché dans le moteur de recherche du guide du tri",
    submitButtonLabel: "Enregistrer les modifications",
    cancelButtonLabel: "Annuler les modifications",
  };
  const maxCharacters = 60;
  const recyclingGuideDisplayContentMaxCharacters = 30;

  /* Methods */
  async function onSubmitValid(submitData: FieldValues) {
    if (submitData["id"]) {
      const variables = {
        updateRecyclingGuideBlockId: submitData["id"],
        data: {
          titleContent: submitData["titleContent"],
          subtitleContent: submitData["subtitleContent"],
          recyclingGuideDisplayContent:
            submitData["recyclingGuideDisplayContent"],
        },
      };
      updateRecyclingGuideBlock({
        variables,
        refetchQueries: [
          {
            query: GetRecyclingBlockTabDocument,
            variables: { contractId },
          },
          "getRecyclingBlockTab",
        ],
      });
      return new Promise<void>((resolve) => {
        setTimeout(() => resolve(), 1000);
      });
    }
  }

  function onCancel() {
    form.reset();
  }

  /* External Data */
  const { contractId } = useContract();
  const { loading, error, data } = useGetRecyclingBlockTabQuery({
    variables: { contractId },
  });
  const [
    updateRecyclingGuideBlock,
    { loading: mutationLoading, error: mutationError },
  ] = useUpdateRecyclingGuideTabMutation();

  /* Local Data */
  const [isShowingSpinner, setIsShowingSpinner] = useState(false);
  const [recyclingGuideData, setRecyclingGuideData] =
    useState<IRecyclingGuideBlock>();
  const formValidationMode = "onChange";
  const form = useForm({
    mode: formValidationMode,
  });
  const { handleSubmit, formState } = form;
  const { isDirty, isSubmitting } = formState;

  useEffect(() => {
    if (data && data.contractCustomizations?.data) {
      const { recyclingGuideBlock } = extractRecyclingGuideBlock(data);
      if (recyclingGuideBlock?.id && recyclingGuideBlock?.attributes) {
        const mappedData = {
          id: recyclingGuideBlock.id,
          titleContent: recyclingGuideBlock.attributes.titleContent,
          subtitleContent: recyclingGuideBlock.attributes.subtitleContent,
          recyclingGuideDisplayContent:
            recyclingGuideBlock.attributes.recyclingGuideDisplayContent,
        };
        setRecyclingGuideData(mappedData);
        form.reset(mappedData);
      }
    }
  }, [form, data]);

  const spinnerTimerRef = useRef<NodeJS.Timeout>();
  useEffect(() => {
    if (isSubmitting) {
      if (!isShowingSpinner) {
        spinnerTimerRef.current = setTimeout(() => {
          setIsShowingSpinner(true);
        }, 200);
      }
    } else {
      clearTimeout(spinnerTimerRef.current);
      setIsShowingSpinner(false);
    }
  }, [isShowingSpinner, isSubmitting]);

  const focusRef = useCallback((node: HTMLFormElement) => {
    FocusFirstElement(node);
  }, []);

  {
    // TODO: layout shift, handle error redirect,
  }
  if (loading) return <CommonSpinner />;
  if (error) return <span>{error?.message}</span>;
  if (mutationError) return <span>{mutationError?.message}</span>;

  return (
    <div
      className="c-RecyclingGuideTab"
      id="panel-recyclingGuide"
      role="tabpanel"
      aria-labelledby="tab-recyclingGuide"
    >
      {isShowingSpinner && <CommonSpinner isCover={true} />}
      <h2 className="c-RecyclingGuideTab__Title">{formLabels.title}</h2>
      <FormProvider {...form}>
        <form
          className="c-RecyclingGuideTab__Form"
          onSubmit={handleSubmit(onSubmitValid)}
          ref={focusRef}
        >
          <div className="c-RecyclingGuideTab__Group c-RecyclingGuideTab__Group_short">
            <FormInput
              type="text"
              name="titleContent"
              label={formLabels.titleContent}
              validationLabel={`${maxCharacters} ${formLabels.maxCharactersLabel}`}
              maxLengthValidation={maxCharacters}
              isRequired={true}
              isDisabled={mutationLoading}
              defaultValue={recyclingGuideData?.titleContent}
            />
          </div>
          <div className="c-RecyclingGuideTab__Group c-RecyclingGuideTab__Group_short">
            <FormInput
              type="text"
              label={formLabels.subtitleContent}
              name="subtitleContent"
              validationLabel={`${maxCharacters} ${formLabels.maxCharactersLabel}`}
              maxLengthValidation={maxCharacters}
              isDisabled={mutationLoading}
              isRequired={true}
              defaultValue={recyclingGuideData?.subtitleContent}
            />
          </div>
          <div className="c-RecyclingGuideTab__Group c-RecyclingGuideTab__Group_short">
            <FormInput
              type="text"
              label={formLabels.recyclingGuideDisplayContent}
              validationLabel={`${recyclingGuideDisplayContentMaxCharacters} ${formLabels.maxCharactersLabel}`}
              name="recyclingGuideDisplayContent"
              isDisabled={mutationLoading}
              isRequired={true}
              maxLengthValidation={recyclingGuideDisplayContentMaxCharacters}
              defaultValue={recyclingGuideData?.recyclingGuideDisplayContent}
            />
          </div>
          <div className="c-RecyclingGuideTab__Buttons">
            <CommonButton
              type="submit"
              label={formLabels.submitButtonLabel}
              style="primary"
              isDisabled={!isDirty}
            />
            <CommonButton
              type="button"
              label={formLabels.cancelButtonLabel}
              onClick={onCancel}
              isDisabled={!isDirty}
            />
          </div>
        </form>
      </FormProvider>
    </div>
  );
}
