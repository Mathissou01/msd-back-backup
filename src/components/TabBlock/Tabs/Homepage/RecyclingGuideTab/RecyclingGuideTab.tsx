import { FormProvider, useForm } from "react-hook-form";
import { FieldValues } from "react-hook-form/dist/types/fields";
import React, { useState, useEffect } from "react";
import {
  useGetRecyclingGuideBlockByContractIdQuery,
  useUpdateRecyclingGuideBlockByIdMutation,
} from "../../../../../graphql/codegen/generated-types";
import { extractRecyclingGuideBlock } from "../../../../../lib/graphql-data";
import { getRightsByLabel } from "../../../../../lib/user";
import { useUser } from "../../../../../hooks/useUser";
import { useContract } from "../../../../../hooks/useContract";
import { useFocusFirstElement } from "../../../../../hooks/useFocusFirstElement";
import CommonLoader from "../../../../Common/CommonLoader/CommonLoader";
import CommonButton from "../../../../Common/CommonButton/CommonButton";
import FormInput from "../../../../Form/FormInput/FormInput";
import "./recycling-guide-tab.scss";

interface IRecyclingGuideBlock {
  id: string;
  titleContent: string;
  subtitleContent: string;
  recyclingGuideDisplayContent: string;
}

export default function RecyclingGuideTab() {
  /* Static Data */
  const formLabels = {
    title: "Guide du Tri",
    maxCharactersLabel: "caractères maximum",
    titleContent: "Titre",
    subtitleContent: "Sous titre",
    recyclingGuideDisplayContent:
      "Texte affiché dans le moteur de recherche du guide du tri",
    submitButtonLabel: "Enregistrer les modifications",
    cancelButtonLabel: "Annuler les modifications",
  };
  const maxCharacters = 60;
  const recyclingGuideDisplayContentMaxCharacters = 30;

  /* Methods */
  async function onSubmitValid(submitData: FieldValues) {
    if (submitData["id"]) {
      const variables = {
        updateRecyclingGuideBlockId: submitData["id"],
        data: {
          titleContent: submitData["titleContent"],
          subtitleContent: submitData["subtitleContent"],
          recyclingGuideDisplayContent:
            submitData["recyclingGuideDisplayContent"],
        },
      };
      return updateRecyclingGuideBlock({
        variables,
      });
    }
  }

  function onCancel() {
    form.reset();
  }

  /* External Data */
  const { contractId } = useContract();
  const { loading, error, data } = useGetRecyclingGuideBlockByContractIdQuery({
    variables: { contractId },
    fetchPolicy: "network-only",
  });
  const [
    updateRecyclingGuideBlock,
    { loading: mutationLoading, error: mutationError },
  ] = useUpdateRecyclingGuideBlockByIdMutation({
    refetchQueries: ["getRecyclingGuideBlockByContractId"],
    awaitRefetchQueries: true,
  });

  /* Local Data */
  const { userRights } = useUser();
  const userPermissions = getRightsByLabel("Homepage", userRights);
  const [recyclingGuideData, setRecyclingGuideData] =
    useState<IRecyclingGuideBlock>();
  const formValidationMode = "onChange";
  const form = useForm({
    mode: formValidationMode,
  });
  const { handleSubmit, formState } = form;
  const { isDirty, isSubmitting } = formState;

  useEffect(() => {
    if (data && data.contractCustomizations?.data) {
      const { recyclingGuideBlock } = extractRecyclingGuideBlock(data);
      if (recyclingGuideBlock?.id && recyclingGuideBlock?.attributes) {
        const mappedData = {
          id: recyclingGuideBlock.id,
          titleContent: recyclingGuideBlock.attributes.titleContent,
          subtitleContent: recyclingGuideBlock.attributes.subtitleContent,
          recyclingGuideDisplayContent:
            recyclingGuideBlock.attributes.recyclingGuideDisplayContent,
        };
        setRecyclingGuideData(mappedData);
        form.reset(mappedData);
      }
    }
  }, [form, data]);

  return (
    <div
      className="c-RecyclingGuideTab"
      id="panel-recyclingGuide"
      role="tabpanel"
      aria-labelledby="tab-recyclingGuide"
    >
      <CommonLoader
        isLoading={loading || isSubmitting || mutationLoading}
        isShowingContent={isSubmitting || mutationLoading}
        hasDelay={isSubmitting || mutationLoading}
        errors={[error, mutationError]}
      >
        <h2 className="c-RecyclingGuideTab__Title">{formLabels.title}</h2>
        <FormProvider {...form}>
          <form
            className="c-RecyclingGuideTab__Form"
            onSubmit={handleSubmit(onSubmitValid)}
            ref={useFocusFirstElement()}
          >
            <div className="c-RecyclingGuideTab__Group c-RecyclingGuideTab__Group_short">
              <FormInput
                type="text"
                name="titleContent"
                label={formLabels.titleContent}
                validationLabel={`${maxCharacters} ${formLabels.maxCharactersLabel}`}
                maxLengthValidation={maxCharacters}
                isRequired={true}
                isDisabled={mutationLoading || !userPermissions.update}
                defaultValue={recyclingGuideData?.titleContent}
              />
            </div>
            <div className="c-RecyclingGuideTab__Group c-RecyclingGuideTab__Group_short">
              <FormInput
                type="text"
                label={formLabels.subtitleContent}
                name="subtitleContent"
                validationLabel={`${maxCharacters} ${formLabels.maxCharactersLabel}`}
                maxLengthValidation={maxCharacters}
                isDisabled={mutationLoading || !userPermissions.update}
                isRequired={true}
                defaultValue={recyclingGuideData?.subtitleContent}
              />
            </div>
            <div className="c-RecyclingGuideTab__Group c-RecyclingGuideTab__Group_short">
              <FormInput
                type="text"
                label={formLabels.recyclingGuideDisplayContent}
                validationLabel={`${recyclingGuideDisplayContentMaxCharacters} ${formLabels.maxCharactersLabel}`}
                name="recyclingGuideDisplayContent"
                isDisabled={mutationLoading || !userPermissions.update}
                isRequired={true}
                maxLengthValidation={recyclingGuideDisplayContentMaxCharacters}
                defaultValue={recyclingGuideData?.recyclingGuideDisplayContent}
              />
            </div>
            <div className="c-RecyclingGuideTab__Buttons">
              <CommonButton
                type="submit"
                label={formLabels.submitButtonLabel}
                style="primary"
                picto="check"
                isDisabled={!isDirty || !userPermissions.update}
              />
              <CommonButton
                type="button"
                label={formLabels.cancelButtonLabel}
                picto="cross"
                onClick={onCancel}
                isDisabled={!isDirty}
              />
            </div>
          </form>
        </FormProvider>
      </CommonLoader>
    </div>
  );
}
