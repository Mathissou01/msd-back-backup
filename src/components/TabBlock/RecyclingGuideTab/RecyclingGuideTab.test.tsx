import { render, screen } from "@testing-library/react";
import { MockedProvider } from "@apollo/client/testing";
import React from "react";
import { defaultMockData } from "../../../../__mocks__/recyclingGuideTabMockData";
import RecyclingGuideTab from "./RecyclingGuideTab";

describe("RecyclingGuideTab", () => {
  it("renders loading and loaded state", async () => {
    const { container } = render(
      <MockedProvider mocks={defaultMockData} addTypename={false}>
        <RecyclingGuideTab />
      </MockedProvider>,
    );

    expect(await screen.findByTestId("common-spinner")).toBeInTheDocument();
    expect(await container).toMatchSnapshot();
    expect(await screen.findByText("Sous titre *")).toBeInTheDocument();
    expect(
      await screen.findByText("30 caractères maximum"),
    ).toBeInTheDocument();
    expect(
      await screen.findByDisplayValue("Titre modifié"),
    ).toBeInTheDocument();
    expect(await container).toMatchSnapshot();
  });
});
