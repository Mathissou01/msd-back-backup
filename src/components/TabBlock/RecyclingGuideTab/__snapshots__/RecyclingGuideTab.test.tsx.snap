// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders loading and loaded state 1`] = `
<div>
  <div
    class="c-CommonSpinner__Wrapper "
  >
    <div
      class="c-CommonSpinner"
      data-testid="common-spinner"
    >
      <div />
      <div />
      <div />
      <div />
    </div>
  </div>
</div>
`;

exports[`renders loading and loaded state 2`] = `
<div>
  <div
    class="c-RecyclingGuideTab"
  >
    <h2
      class="c-RecyclingGuideTab__Title"
    >
      Guide du tri
    </h2>
    <form
      class="c-RecyclingGuideTab__Form"
    >
      <div
        class="c-RecyclingGuideTab__Group"
      >
        <div
          class="c-FormInput"
        >
          <label
            class="c-FormLabel"
            for="titleContent"
          >
            <div
              class="c-FormLabel__Label"
            >
              <span>
                Titre *
              </span>
            </div>
            <span
              class="c-FormLabel__Validation"
            >
              60 caractères maximum
            </span>
          </label>
          <input
            aria-errormessage="titleContent_error"
            aria-invalid="false"
            class="c-FormInput__Input "
            data-testid="form-input"
            id="titleContent"
            maxlength="60"
            name="titleContent"
            type="text"
            value="Titre modifié"
          />
        </div>
      </div>
      <div
        class="c-RecyclingGuideTab__Group"
      >
        <div
          class="c-FormInput"
        >
          <label
            class="c-FormLabel"
            for="subtitleContent"
          >
            <div
              class="c-FormLabel__Label"
            >
              <span>
                Sous titre *
              </span>
            </div>
            <span
              class="c-FormLabel__Validation"
            >
              60 caractères maximum
            </span>
          </label>
          <input
            aria-errormessage="subtitleContent_error"
            aria-invalid="false"
            class="c-FormInput__Input "
            data-testid="form-input"
            id="subtitleContent"
            maxlength="60"
            name="subtitleContent"
            type="text"
            value="Trouver les consignes de tri dans ma ville"
          />
        </div>
      </div>
      <div
        class="c-RecyclingGuideTab__Group"
      >
        <div
          class="c-FormInput"
        >
          <label
            class="c-FormLabel"
            for="recyclingGuideDisplayContent"
          >
            <div
              class="c-FormLabel__Label"
            >
              <span>
                Texte affiché dans le moteur de recherche du guide du tri *
              </span>
            </div>
            <span
              class="c-FormLabel__Validation"
            >
              30 caractères maximum
            </span>
          </label>
          <input
            aria-errormessage="recyclingGuideDisplayContent_error"
            aria-invalid="false"
            class="c-FormInput__Input "
            data-testid="form-input"
            id="recyclingGuideDisplayContent"
            maxlength="30"
            name="recyclingGuideDisplayContent"
            placeholder="Lampe, écran, épluchure..."
            type="text"
            value="Lampe, écran, épluchure"
          />
        </div>
      </div>
      <div
        class="c-RecyclingGuideTab__Buttons"
      >
        <button
          class="c-CommonButton c-CommonButton_primary"
          disabled=""
          type="submit"
        >
          Enregistrer les modifications
        </button>
        <button
          class="c-CommonButton c-CommonButton_secondary"
          disabled=""
          type="button"
        >
          Annuler les modifications
        </button>
      </div>
    </form>
  </div>
</div>
`;
